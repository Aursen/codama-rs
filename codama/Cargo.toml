[package]
name = "codama"
version = "0.1.0"
edition = "2021"

[dependencies]
codama-errors = { path = "../codama-errors" }
codama-korok-plugins = { path = "../codama-korok-plugins" }
codama-korok-visitors = { path = "../codama-korok-visitors" }
codama-koroks = { path = "../codama-koroks" }
codama-macros = { path = "../codama-macros" }
codama-nodes = { path = "../codama-nodes" }
codama-stores = { path = "../codama-stores" }
proc-macro2 = "1.0"
