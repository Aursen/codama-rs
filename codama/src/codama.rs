use codama_errors::{CodamaError, CodamaResult};
use codama_korok_plugins::{resolve_plugins, DefaultPlugin, KorokPlugin};
use codama_koroks::RootKorok;
use codama_nodes::{Node, NodeTrait, NodeUnionTrait, RootNode};
use codama_stores::RootStore;
use std::path::Path;

pub struct Codama {
    store: RootStore,
    plugins: Vec<Box<dyn KorokPlugin>>,
}

impl Codama {
    pub fn new(store: RootStore) -> Self {
        Self {
            store,
            plugins: Vec::new(),
        }
        .add_plugin(DefaultPlugin)
    }

    pub fn load(path: &Path) -> CodamaResult<Self> {
        Ok(Self::new(RootStore::load(path)?))
    }

    pub fn load_all(paths: &[&Path]) -> CodamaResult<Self> {
        Ok(Self::new(RootStore::load_all(paths)?))
    }

    pub fn hydrate(tt: proc_macro2::TokenStream) -> CodamaResult<Self> {
        Ok(Self::new(RootStore::hydrate(tt)?))
    }

    pub fn add_plugin<T: KorokPlugin + 'static>(mut self, plugin: T) -> Self {
        self.plugins.push(Box::new(plugin));
        self
    }

    pub fn get_korok(&self) -> CodamaResult<RootKorok> {
        RootKorok::parse(&self.store)
    }

    pub fn get_visited_korok(&self) -> CodamaResult<RootKorok> {
        let mut korok = self.get_korok()?;
        let run_plugins = resolve_plugins(&self.plugins);
        run_plugins(&mut korok);
        Ok(korok)
    }

    pub fn get_node(&self) -> CodamaResult<Node> {
        let korok = self.get_visited_korok()?;
        korok.node.clone().ok_or(CodamaError::NodeNotFound)
    }

    pub fn get_idl(&self) -> CodamaResult<RootNode> {
        let node = self.get_node()?;
        match node {
            Node::Root(root) => Ok(root),
            _ => Err(CodamaError::UnexpectedNode {
                expected: "RootNode".to_string(),
                actual: node.kind().to_string(),
            }),
        }
    }

    pub fn get_json_idl(&self) -> CodamaResult<String> {
        self.get_idl()?.to_json()
    }
}
