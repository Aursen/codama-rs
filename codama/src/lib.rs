mod codama;

pub use codama::*;
pub use codama_errors::*;
pub use codama_korok_plugins::*;
pub use codama_korok_visitors::*;
pub use codama_koroks::*;
pub use codama_nodes::*;
pub use codama_stores::*;
