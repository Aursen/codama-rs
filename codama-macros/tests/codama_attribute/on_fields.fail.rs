use codama::{codama, CodamaType};

#[derive(CodamaType)]
pub struct StructTest {
    #[codama(node())]
    pub public_field: u8,
    #[codama(node())]
    private_field: u8,
}

#[derive(CodamaType)]
pub struct TupleTest(#[codama(node())] u8);

#[derive(CodamaType)]
pub enum EnumTest {
    Struct {
        #[codama(node())]
        field: u8,
    },
    Tuple(#[codama(node())] u8),
}

fn main() {}
