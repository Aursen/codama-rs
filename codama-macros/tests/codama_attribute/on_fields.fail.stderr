error: unrecognized type
 --> tests/codama_attribute/on_fields.fail.rs:5:21
  |
5 |     #[codama(type = invalid)]
  |                     ^^^^^^^

error: unrecognized type
  --> tests/codama_attribute/on_fields.fail.rs:12:38
   |
12 | pub struct TupleTest(#[codama(type = invalid)] u8);
   |                                      ^^^^^^^

error: unrecognized type
  --> tests/codama_attribute/on_fields.fail.rs:17:25
   |
17 |         #[codama(type = invalid)]
   |                         ^^^^^^^
